src/Microsoft.IdentityModel.Tokens/Encryption/Cng.cs

```diff
@@ -0,0 +1,216 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Security.Cryptography;
+using Microsoft.IdentityModel.Logging;
+using static Microsoft.IdentityModel.Tokens.Interop.BCrypt;
+
+namespace Microsoft.IdentityModel.Tokens
+{
+    internal static class CryptoThrowHelper
+    {
+        public static CryptographicException ToCryptographicException(this int hr)
+        {
+            string message = Interop.Kernel32.GetMessage(hr);
+
+            if ((hr & 0x80000000) != 0x80000000)
+                hr = (hr & 0x0000FFFF) | unchecked((int)0x80070000);
+
+            return new WindowsCryptographicException(hr, message);
+        }
+
+        private sealed class WindowsCryptographicException : CryptographicException
+        {
+            public WindowsCryptographicException(int hr, string message)
+                : base(message)
+            {
+                HResult = hr;
+            }
+
+            public WindowsCryptographicException(string message) : base(message)
+            {
+            }
+
+            public WindowsCryptographicException(string message, Exception innerException) : base(message, innerException)
+            {
+            }
+
+            public WindowsCryptographicException()
+            {
+            }
+        }
+    }
+
+    internal static class AesBCryptModes
```

Comment:
Is this code vendored from another source?

Reply:
Yes, from [.NET core](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography)

Reply:
We should try to make the clear with comments. We don't want to change this code other than potentially resyncing with the source in the future.

Reply:
@RojaEnnam @keegan-caruso we want to make sure we throw (as close as possible) the same exceptions so that our users will not see any difference between Core and .Net Desktop.

Reply:
@RojaEnnam Thoughts on grouping vendored code under a specific directory?

Reply:
I think its a good idea. There is a opensource folder already. I've moved the Aesgcm related vendor code to it.